go/libraries/utils/svcs/controller.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcs
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// A Service is a runnable unit of functionality that a Controller can
+// take responsibility for.  It has an |Init| function, which can error, and
+// which should do all of the initialization and validation work necessary to
+// bring the service up. It has a |Run| function, which will be called in a
+// separate go-routine and should run and provide the functionality associated
+// with the service until the |Stop| function is called.
+type Service struct {
+	Init func(context.Context) error
+	Run  func(context.Context)
+	Stop func() error
+}
+
+// A Controller is responsible for initializing a number of registered
+// services, running them all, and stopping them all when requested. Services
+// are registered with |Register(*Service)|. When |Start| is called, the
+// services are all initialized,  in the order of their registration, and if
+// every service initializes successfully, they are |Run| concurrently. When
+// |Stop| is called, services are stopped in reverse-registration order. |Stop|
+// does not block for the goroutines spawned by |Start| to complete, although
```

Comment:
Probably document the fact that |Stop| *does* block until Start is done calling Stop() on all services.